### PR TITLE
Font scaling options

### PR DESCRIPTION
**Pull request by celiasmith**
This allows the user to turn on/off scaling of fonts with zoom.  It also allows defining a % of base for the font size.  This addresses #113

### CODE REVIEW

**Comment by jgosmann** [comment]
* [ ] Can't select text in the font size input box in Firefox.
* [ ] I think, it could be more useful to have 10% or 20% steps when clicking the arrows
* [ ] If I put the cursor into the text box and press enter, the state of 'allow fonts to zoom' is lost.

Do we need that input box at all? The browser should provide the same functionality. The only difference is that the browser scales also the top and bottom toolbar, but we could set a fixed font-size for those.

File: nengo_viz/components/sim_control.py [hunk 1]
```diff
@@ -86,9 +86,11 @@ def update_client(self, client):
             client.write('reload')
             self.reload = False
         if self.send_config_options == True:
-            from ..disposable_js import configmodal
-            javascript = configmodal()
-            client.write('config' + javascript)
+            options = ["zoom"] #This is ignored by the client for now
+                               #But will be useful when saving/loading options
+            js_options = json.dumps(options)
```
Comment (jgosmann) [inline nengo_viz/components/sim_control.py]: I would prefer to exclude such code in the merge (we can keep it in a separate branch).

**Comment by jgosmann** [comment]
Also wondering, would it be better if one can set the base font size in pt because it's a more common unit?

File: nengo_viz/static/viz_netgraph.js [hunk 2]
```diff
@@ -8,9 +8,11 @@
  */
 VIZ.NetGraph = function(parent, args) {
     this.scale = 1.0;          // global scaling factor
-    this.offsetX = 0;          // global x,y pan offset 
+    this.offsetX = 0;          // global x,y pan offset
     this.offsetY = 0;
-    
+    this.zoom_fonts = false;    // scale fonts when zooming
+    this.font_size = 100;       // font size as a percent of base
```
Comment (jgosmann) [inline nengo_viz/static/viz_netgraph.js]: I would prefer to store the actual fraction, i.e. `font_size` divided by 100. Then it is not necessary to remember the scaling factor whenever this value is used (only for the input form it has to be converted to a percentage).

File: nengo_viz/static/viz_netgraph.js [hunk 3]
```diff
@@ -298,22 +300,40 @@ VIZ.NetGraph.prototype.set_scale = function(scale) {
 
 
 VIZ.NetGraph.prototype.update_font_size = function(scale) {
-    // TODO: text resizing is disabled for now until we can figure out
-    // how best to have it work with large models
+    if (this.zoom_fonts) {
+        $('#main').css('font-size', 3 * this.scale * this.font_size/100 + 'em');
```
Comment (jgosmann) [inline nengo_viz/static/viz_netgraph.js]: Do we want to keep that factor of 3 or should we make the font-size match independent of the zoom mode on scale = 1.

File: nengo_viz/static/viz_top_toolbar.css [hunk 4]
```diff
@@ -68,3 +68,24 @@
     margin-top: 3em;
     position: absolute;
 }
+
+#config-fontsize-group {
+    margin-left:20;
```
Comment (jgosmann) [inline nengo_viz/static/viz_top_toolbar.css]: Missing a unit ...

File: nengo_viz/static/viz_top_toolbar.css [hunk 5]
```diff
@@ -68,3 +68,24 @@
...
+}
+
+#config-fontsize.form-control {
+    padding: 6px 6px;
+    height: 35px;
```
Comment (jgosmann) [inline nengo_viz/static/viz_top_toolbar.css]: It might be better to use `em` here.

File: nengo_viz/static/viz_top_toolbar.css [hunk 6]
```diff
@@ -68,3 +68,24 @@
...
+}
+
+#config-fontsize-group .control-label {
+    margin-left: -20px;
```
Comment (jgosmann) [inline nengo_viz/static/viz_top_toolbar.css]: Is this undoing the effect of line 73? Maybe both should be deleted?

File: nengo_viz/static/viz_top_toolbar.js [hunk 7]
```diff
@@ -82,8 +82,33 @@ VIZ.Toolbar.prototype.reset_model_layout = function () {
...
-VIZ.Toolbar.prototype.start_modal = function () {
-    sim.ws.send('config')
+ *  call to server to call config_modal_show with config data. */
+VIZ.Toolbar.prototype.config_modal = function () {
+    sim.ws.send('config');  //Doing it this way in case we need to save options to a file later
+}
+
+VIZ.Toolbar.prototype.config_modal_show = function(options) {
+    var self = this;
+    console.log(options); //Options are ignored for now
```
Comment (jgosmann) [inline nengo_viz/static/viz_top_toolbar.js]: I think, this should be removed because it's just debug output.

**Comment by jgosmann** [comment]
* [ ] The clickable arrows are pretty small (hard to hit).
* [ ] The spacing between the input box and the 'allow font zoom' checkbox feels too large. These are related options, but they space makes them look unrelated.

**Comment by jgosmann** [comment]
We might no need to fix all of that before the summer school, though.

File: nengo_viz/static/viz_top_toolbar.css (hunk 5 above)
Comment (celiasmith) [inline nengo_viz/static/viz_top_toolbar.css]: bootstrap defines these in px

File: nengo_viz/components/sim_control.py (hunk 1 above)
Comment (celiasmith) [inline nengo_viz/components/sim_control.py]: done

File: nengo_viz/static/viz_netgraph.js (hunk 2 above)
Comment (celiasmith) [inline nengo_viz/static/viz_netgraph.js]: Seems to me this adds unnecessary complexity (stripping decimals, etc.), and it's a documented unit.

File: nengo_viz/static/viz_netgraph.js (hunk 3 above)
Comment (celiasmith) [inline nengo_viz/static/viz_netgraph.js]: I'm fine either way... in both cases we'll be putting parameters in somewhere (either here or various .css files).

File: nengo_viz/static/viz_top_toolbar.css (hunk 4 above)
Comment (celiasmith) [inline nengo_viz/static/viz_top_toolbar.css]: fixed

File: nengo_viz/static/viz_top_toolbar.css (hunk 6 above)
Comment (celiasmith) [inline nengo_viz/static/viz_top_toolbar.css]: No it isn't. This is positioning the label and the text input box separately.

**Comment by celiasmith** [comment]
> The spacing between the input box and the 'allow font zoom' checkbox feels too large. These are related options, but they space makes them look unrelated.

I agree.  I was a bit torn about playing with spacing too much, since right now it's consistent between all modals.  These are defined largely by the bootstrap defaults atm.

**Comment by tcstewar** [comment]
> Do we need that input box at all? The browser should provide the same functionality. The only difference is that the browser scales also the top and bottom toolbar, but we could set a fixed font-size for those.

There'll also generally be a difference between the font size for the NetGraph and the font size for the code editor....  In presentations I often want to make the code font bigger, but not the NetGraph.

**Comment by celiasmith** [comment]
> The clickable arrows are pretty small (hard to hit).

This can only be changed on chrome it seems.  So, i'll just leave it to be consistent across browsers.

**Comment by celiasmith** [comment]
> can we get the 'as percentage' text right of the input field (instead of weirdly aligned below)?

This is the bootstrap/validator default alignment for such text (and for errors).  So, i'll leave it to be consistent with all the other modals.

File: nengo_viz/static/viz_netgraph.js (hunk 2 above)
Comment (jgosmann) [inline nengo_viz/static/viz_netgraph.js]: [Here the unit was also documented and it still went wrong.](https://en.wikipedia.org/wiki/Mars_Climate_Orbiter) Good thing we're not building space crafts. :rocket: :boom:

File: nengo_viz/static/viz_top_toolbar.css (hunk 5 above)
Comment (jgosmann) [inline nengo_viz/static/viz_top_toolbar.css]: If bootstrap defines them, why do we have to redefine them?

**Comment by celiasmith** [comment]
> Can't select text in the font size input box in Firefox.

This seems to be some weird firefox bug, and i can't track it down.  You can't select anything in the modals with the mouse (the keyboard works fine).

**Comment by tbekolay** [comment]
I don't think this is a firefox bug, I think we deliberately set everything to be unselectable? There's something in the CSS about not being able to select text... yeah. [Here it is](https://github.com/nengo/nengo_viz/blob/cd3edbb381b4b8d1a444de1743b3523e837549db/nengo_viz/static/main.css#L29). This should probably be more specific.

File: nengo_viz/static/viz_netgraph.js (hunk 2 above)
Comment (celiasmith) [inline nengo_viz/static/viz_netgraph.js]: guess we shouldn't have units... no, wait a sec :stuck_out_tongue_winking_eye:

**Comment by celiasmith** [comment]
> I think we deliberately set everything to be unselectable?

I guess chrome is ignoring us then... and so is firefox when we use the keyboard?  Soooo confused :confused: ... I'll reset the css for this element and see.

**Comment by celiasmith** [comment]
No luck with the css... styling any id, class, or tag seems to style the entire document, so either it's full of annoying text selecting mouse cursors in netgraph, or you can't select text with the mouse (in firefox and safari).  I think the latter is the best of these bad options.

I think this is ready to go unless there are other outstanding issues i'm forgetting.

**Comment by jgosmann** [comment]
[That CSS property seems to be non-standard with a big warning banner.](https://developer.mozilla.org/en-US/docs/Web/CSS/user-select) Makes me wonder whether we should be using it at all? And it seems that Firefox can re-enable selection only with `-moz-user-select:text`.

**Comment by jgosmann** [comment]
Moreover, `user-select` does not change the cursor in Firefox, so we actually should set the cursor to `default` wherever we have non-selectable text, but in a way which does not break the resize cursors.

**Comment by celiasmith** [comment]
I created another branch (text-selecting) to look at fixing the selection stuff, as that's beyond what this is about.

**Comment by celiasmith** [comment]
PR #284 takes care of the text selection problem in the modal.

File: nengo_viz/static/viz.css [hunk 8]
```diff
@@ -45,4 +45,4 @@
 
 .graph .label:active {
     cursor: ns-resize;
-}
\ No newline at end of file
+}
```
Comment (tbekolay) [inline nengo_viz/static/viz.css]: Looks like this is only whitespace changes... amend this out?

File: nengo_viz/static/viz_netgraph.css [hunk 9]
```diff
@@ -41,5 +41,5 @@ g.net rect{
 .netgraph g.passthrough text {
     text-anchor: middle;
     dominant-baseline: text-before-edge;
-    }    
+    }
```
Comment (tbekolay) [inline nengo_viz/static/viz_netgraph.css]: Also just whitespace changes in this file.

File: nengo_viz/static/viz_top_toolbar.css [hunk 10]
```diff
@@ -68,3 +68,25 @@
...
+#config-fontsize.form-control {
+    padding: 6px 6px;
+    height: 35px;
+}
+
+.input-group .input-group-addon {
+    text-indent:0px;
+}
+
+#myModalForm .form-group {
```
Comment (tbekolay) [inline nengo_viz/static/viz_top_toolbar.css]: These should probably be in `viz_modal.css`?

Comment (celiasmith) [inline nengo_viz/static/viz_top_toolbar.css]: done!

**Comment by tbekolay** [comment]
Tried this out. Issues I had:

- I tried to make the font size 125%, but that failed validation. It would only let me put in multiples of 10. Feels weird, should allow any number greater than 0...
- There's a text input box here that's not connected to anything? I think you're missing a call to `VIZ.modal.clear_body` in the config body function.

The actual text scaling checkbox works perfectly though ;) It would be kind of nice if it updated the model when you click the checkbox, but definitely not essential.

**Comment by celiasmith** [comment]
> I tried to make the font size 125%, but that failed validation

It's not clear how to do this, I've submitted an issue to the validator github.  I can't override the default validator.  So, either I turn off the clickable by 10, or you have to do steps of 10.  I've left it for now.

The text input was just a dumb mistake :unamused: ... instant update for another day... (it should update immediately on exit from the modal).

**Comment by celiasmith** [comment]
BTW, the developer responded: 

"This isn't really possible with the way the validator is currently written. The only viable solution at the moment is to add an option like data-disable-native=true, but since the native validators are all bundled together, you'd lose them all (required/min/max/etc.). I guess that's not a huge issue since you'll be adding a custom validator and can check whatever you need to in there, but it's kinda coupling and not ideal."

I actually figured out how to disable the native validator, but then you have to write validators for *all* classes.  I thought about doing this because I only needed to add the checkbox, but this will be crazy if the options expand.

**Comment by tbekolay** [comment]
> data-disable-native=true

Yeah... I think it's maybe fine to just have it update in steps on 1? Can't people just click and hold? Can we change the rate at which it gets bigger when you hold?

**Comment by celiasmith** [comment]
So I implemented your other suggest (immediate apply changes)... good idea!  

Yes, if you hold, it will spin (I don't think you can change the speed).  I'm thinking steps of 5 maybe?

**Comment by celiasmith** [comment]
Unfortunately the immediate update doesn't work if you hold down the arrows... so this makes steps of 5 seem good, since you can see the change on each click (can't really with 1).

**Comment by celiasmith** [comment]
OK, I fixed the immediate update to now work when you hold the button down.  So, I also changed the step size to one, so you can watch it grow as you hold the button down.

**Comment by jgosmann** [comment]
If we keep the immediate update, the `Cancel` button should be removed, because it doesn't undo the changes.

**Comment by jgosmann** [comment]
I lied ... if there's a validation error only `Cancel` allows to close the window. But that doesn't change that he button does not have the expected effect.

Otherwise, this looks pretty good now. :sparkles:

**Comment by tbekolay** [comment]
This LGTM. Any objections to me cleaning up the history and merging?

**Comment by celiasmith** [comment]
I just made the change jan suggested of actually resetting all properties on cancel (despite immediate update).  So... a quick review of that, and this should be ready to go.

**Comment by celiasmith** [comment]
I was going to say, it won't work with !==, but does with !=, but you're right, being more explicit 'undefined' is better....

**Comment by tcstewar** [comment]
> I was going to say, it won't work with !==, but does with !=, but you're right, being more explicit 'undefined' is better....

I think that's because it's supposed to be 

```
if (cancel_function !== undefined)
```

```undefined``` is a keyword in javascript.

**Comment by jgosmann** [comment]
[How to check for undefined.](https://stackoverflow.com/questions/3390396/how-to-check-for-undefined-in-javascript)

**Comment by tbekolay** [comment]
Okay cool; I'm going to rebase and fix up the history, then we can do a final lookover before merging.

**Comment by tbekolay** [comment]
OK, rebased and stripped out whitespace changes. If there are no weird unforeseen interactions with new stuff in master, this should be good to go!

**Pull request merged**